extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitter.java

```diff
@@ -0,0 +1,153 @@
+  {
+    return new PrometheusEmitter(config);
+  }
+
+  public PrometheusEmitter(PrometheusEmitterConfig config)
+  {
+    this.config = config;
+    this.strategy = config.getStrategy();
+    metrics = new Metrics(config.getNamespace(), config.getDimensionMapPath());
+  }
+
+
+  @Override
+  public void start()
+  {
+    if (strategy.equals(PrometheusEmitterConfig.Strategy.exporter)) {
+      if (server == null) {
+        try {
+          server = new HTTPServer(config.getPort());
+        }
+        catch (IOException e) {
+          log.error(e, "Unable to start prometheus HTTPServer");
+        }
+      } else {
+        log.error("HTTPServer is already started");
+      }
+    } else if (strategy.equals(PrometheusEmitterConfig.Strategy.pushgateway)) {
+      pushGateway = new PushGateway(config.getPushGatewayAddress());
+    }
+
+  }
+
+  @Override
+  public void emit(Event event)
+  {
+    if (event instanceof ServiceMetricEvent) {
+      emitMetric((ServiceMetricEvent) event);
+    }
+  }
+
+  void emitMetric(ServiceMetricEvent metricEvent)
+  {
+    String name = metricEvent.getMetric();
+    String service = metricEvent.getService();
+    Map<String, Object> userDims = metricEvent.getUserDims();
+    identifier = (userDims.get("task") == null ? metricEvent.getHost() : (String) userDims.get("task"));
+    Number value = metricEvent.getValue();
+
+    DimensionsAndCollector metric = metrics.getByName(name, service);
+    if (metric != null) {
+      String[] labelValues = new String[metric.getDimensions().length];
+      String[] labelNames = metric.getDimensions();
+      for (int i = 0; i < labelValues.length; i++) {
+        String labelName = labelNames[i];
+        //labelName is controlled by the user. Instead of potential NPE on invalid labelName we use "unknown" as the dimension value
+        Object userDim = userDims.get(labelName);
+        labelValues[i] = userDim != null ? pattern.matcher(userDim.toString()).replaceAll("_") : "unknown";
+      }
+
+      if (metric.getCollector() instanceof Counter) {
+        ((Counter) metric.getCollector()).labels(labelValues).inc(value.doubleValue());
+      } else if (metric.getCollector() instanceof Gauge) {
+        ((Gauge) metric.getCollector()).labels(labelValues).set(value.doubleValue());
+      } else if (metric.getCollector() instanceof Histogram) {
+        ((Histogram) metric.getCollector()).labels(labelValues).observe(value.doubleValue() / metric.getConversionFactor());
+      } else {
+        log.error("Unrecognized metric type [%s]", metric.getCollector().getClass());
+      }
+    } else {
+      log.debug("Unmapped metric [%s]", name);
+    }
+  }
+
+  @Override
+  public void flush()
+  {
+    Map<String, DimensionsAndCollector> map = metrics.getMap();
+    try {
+      for (DimensionsAndCollector collector : map.values()) {
+        pushGateway.push(collector.getCollector(), config.getNamespace(), ImmutableMap.of(config.getNamespace(), identifier));
```

Comment:
potential NPE? if the configured strategy is not pushgateway, then this `pushGateway` wouldn't have been instantiated

Reply:
Also should we use a more meaningful label name for `identifier` instead of using the `config.getNamespace()`?

Reply:
Will add the null check - however flush() for this emitter should only called by close() which strategy check already done.

Reply:
For the identifier label name, any suggestion? The `config.namespace` will be set in config files for each service. So for example peon task it would be peon=taskXXX as groupingKey.